src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -0,0 +1,525 @@
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("mediaId must be a non-empty string"),
+		}
+	}
+	if r.MediaMetadata.Origin == "" {
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound("serverName must be a non-empty string"),
+		}
+	}
+	return nil
+}
+
+func (r *downloadRequest) jsonErrorResponse(w http.ResponseWriter, res util.JSONResponse) {
+	// Marshal JSON response into raw bytes to send as the HTTP body
+	resBytes, err := json.Marshal(res.JSON)
+	if err != nil {
+		r.Logger.WithError(err).Error("Failed to marshal JSONResponse")
+		// this should never fail to be marshalled so drop err to the floor
+		res = util.MessageResponse(500, "Internal Server Error")
+		resBytes, _ = json.Marshal(res.JSON)
+	}
+
+	// Set status code and write the body
+	w.WriteHeader(res.Code)
+	r.Logger.WithField("code", res.Code).Infof("Responding (%d bytes)", len(resBytes))
+	w.Write(resBytes)
+}
+
+var nTries = 5
+
+// Download implements /download
+// Files from this server (i.e. origin == cfg.ServerName) are served directly
+// Files from remote servers (i.e. origin != cfg.ServerName) are cached locally.
+// If they are present in the cache, they are served directly.
+// If they are not present in the cache, they are obtained from the remote server and
+// simultaneously served back to the client and written into the cache.
+func Download(w http.ResponseWriter, req *http.Request, origin gomatrixserverlib.ServerName, mediaID types.MediaID, cfg *config.MediaAPI, db *storage.Database, activeRemoteRequests *types.ActiveRemoteRequests) {
+	r := &downloadRequest{
+		MediaMetadata: &types.MediaMetadata{
+			MediaID: mediaID,
+			Origin:  origin,
+		},
+		Logger: util.GetLogger(req.Context()),
+	}
+
+	// request validation
+	if req.Method != "GET" {
+		r.jsonErrorResponse(w, util.JSONResponse{
+			Code: 405,
+			JSON: jsonerror.Unknown("request method must be GET"),
+		})
+		return
+	}
+
+	if resErr := r.Validate(); resErr != nil {
+		r.jsonErrorResponse(w, *resErr)
+		return
+	}
+
+	// check if we have a record of the media in our database
+	mediaMetadata, err := db.GetMediaMetadata(r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+
+	if err == nil {
+		// If we have a record, we can respond from the local file
+		r.MediaMetadata = mediaMetadata
+		r.respondFromLocalFile(w, cfg.AbsBasePath)
+		return
+	} else if err == sql.ErrNoRows && r.MediaMetadata.Origin != cfg.ServerName {
+		// If we do not have a record and the origin is remote, we need to fetch it and respond with that file
+		// The following code using activeRemoteRequests is avoiding duplication of fetches from the remote server in the case
+		// of multiple simultaneous incoming requests for the same remote file - it will be downloaded once, cached and served
+		// to all clients.
+
+		mxcURL := "mxc://" + string(r.MediaMetadata.Origin) + "/" + string(r.MediaMetadata.MediaID)
+
+		for tries := 0; ; tries++ {
+			activeRemoteRequests.Lock()
```

Comment:
I have no confidence that this will always be unlocked. Please convince me otherwise. What does the for loop do? It might be easier to split this into a smaller function and use `defer` at appropriate points.